prebuild/emscripten/post.js

```diff
@@ -0,0 +1,1010 @@
+  var self = this.ptr;
+  const [_scalar, scalar_len] = ensureRef(scalar);
+  return wrapPointer(_emscripten_bind_P2_mult_1(self, _scalar, scalar_len * 8), P2);
+};;
+
+P2.prototype['neg'] = P2.prototype.neg = /** @suppress {undefinedVars, duplicate} @this{Object} */function() {
+  var self = this.ptr;
+  return wrapPointer(_emscripten_bind_P2_neg_0(self), P2);
+};;
+
+P2.prototype['cneg'] = P2.prototype.cneg = /** @suppress {undefinedVars, duplicate} @this{Object} */function(flag) {
+  var self = this.ptr;
+  if (flag && typeof flag === 'object') flag = flag.ptr;
+  return wrapPointer(_emscripten_bind_P2_cneg_1(self, flag), P2);
+};;
+
+P2.prototype['add'] = P2.prototype.add = /** @suppress {undefinedVars, duplicate} @this{Object} */function(a) {
+  var self = this.ptr;
+  switch(a.constructor) {
+    case P2:
+      return wrapPointer(_emscripten_bind_P2_add_1__P2(self, a.ptr), P2);
+    case P2_Affine:
+      return wrapPointer(_emscripten_bind_P2_add_1__P2_Affine(self, a.ptr), P2);
+    default:
+      throw new Error("unsuported type for `a` arg");
+  }
+};;
+
+P2.prototype['dbl'] = P2.prototype.dbl = /** @suppress {undefinedVars, duplicate} @this{Object} */function() {
+  var self = this.ptr;
+  return wrapPointer(_emscripten_bind_P2_dbl_0(self), P2);
+};;
+
+  P2.prototype['__destroy__'] = P2.prototype.__destroy__ = /** @suppress {undefinedVars, duplicate} @this{Object} */function() {
+  var self = this.ptr;
+  _emscripten_bind_P2___destroy___0(self);
+};
+// SecretKey
+/** @suppress {undefinedVars, duplicate} @this{Object} */function SecretKey() {
+  this.ptr = _emscripten_bind_SecretKey_SecretKey_0();
+  getCache(SecretKey)[this.ptr] = this;
+};;
+SecretKey.prototype = Object.create(WrapperObject.prototype);
+SecretKey.prototype.constructor = SecretKey;
+SecretKey.prototype.__class__ = SecretKey;
+SecretKey.__cache__ = {};
+Module['SecretKey'] = SecretKey;
+
+SecretKey.prototype['keygen'] = SecretKey.prototype.keygen = /** @suppress {undefinedVars, duplicate} @this{Object} */function(ikm, info) {
+  var self = this.ptr;
+  var ikm_len = 0;
+  ensureCache.prepare();
+  if (ikm && typeof ikm === 'object') {
+    if (ikm instanceof Uint8Array) {
+      ikm_len = ikm.length;
+      ikm = ensureUint8(ikm, ikm.length);
+    } else {
+      throw new Error("unsupported ikm type");
+    }
+  } else {
+    ikm_len = ikm.length;
+    ikm = ensureString(ikm);
+  }
+  if (info && typeof info === 'object') info = info.ptr;
+  else info = ensureString(info || "");
+  _emscripten_bind_SecretKey_keygen_3(self, ikm, ikm_len, info);
+};;
+
+SecretKey.prototype['from_bendian'] = SecretKey.prototype.from_bendian = /** @suppress {undefinedVars, duplicate} @this{Object} */function(input) {
+  var self = this.ptr;
+  ensureCache.prepare();
+  if (typeof input == 'object') { input = ensureInt8(input); }
+  _emscripten_bind_SecretKey_from_bendian_1(self, input);
```

Comment:
Who deallocates `input`? BTW, why is it copied to stack in the bind method? Is it actually required? BTW, I would argue that a buffer length check would be appropriate here.

Reply:
I'm used to believing that webasm can't access the Javascript pool. But now I see things about `emscripten::val` and it looks like one can create a memoryView of webasm memory in C++ and call its "set" method to copy data from the Javascript pool. Do you have any experience with that? Idea would be that instead of allocating `input` from Javascript one would pass the Uint8Array Javascript object, have C++ copy data to stack and deserialize it into SecretKey. Same technique can be applied to all deserialization methods...

Reply:
> `emscripten::val`

It doesn't seem to be possible to pass them as arguments.